klippa/src/glyf_loca.rs

```diff
@@ -0,0 +1,298 @@
+                let trimmed_len = subset_glyph.len();
+                max_offset += padded_size(trimmed_len) as u32;
+                subset_glyphs.push(subset_glyph);
+            }
+            _ => {
+                return Err(SubsetTableError(Glyf::TAG));
+            }
+        }
+    }
+
+    //TODO: support force_long_loca in the plan
+    let loca_format: u8 = if max_offset < 0x1FFFF { 0 } else { 1 };
+
+    let glyf_cap = estimate_subset_table_size(font, Glyf::TAG, plan);
+    let mut glyf_out = Vec::with_capacity(glyf_cap);
+
+    let loca_cap = estimate_subset_table_size(font, Loca::TAG, plan);
+    let mut loca_out: Vec<u8> = Vec::with_capacity(loca_cap);
+
+    if loca_format == 0 {
+        loca_out.extend_from_slice(&0_u16.to_be_bytes());
+        let mut offset: u16 = 0;
+        for g in &subset_glyphs {
+            let padded_len = padded_size(g.len());
+            offset += padded_len as u16;
+            let glyph_offset = offset >> 1;
+            loca_out.extend_from_slice(&glyph_offset.to_be_bytes());
+            glyf_out.extend_from_slice(g);
+            if padded_len > g.len() {
+                glyf_out.extend_from_slice(&[0]);
+            }
+        }
+    } else {
+        loca_out.extend_from_slice(&0_u32.to_be_bytes());
+        let mut offset: u32 = 0;
+        for g in &subset_glyphs {
+            offset += g.len() as u32;
+            loca_out.extend_from_slice(&offset.to_be_bytes());
+            glyf_out.extend_from_slice(g);
+        }
+    }
+
+    // As a special case when all glyph in the font are empty, add a zero byte to the table,
+    // so that OTS doesn’t reject it, and to make the table work on Windows as well.
+    // See https://github.com/khaledhosny/ots/issues/52
+    if glyf_out.is_empty() {
+        glyf_out.extend_from_slice(&[0]);
+    }
+
+    let Ok(head_out) = subset_head(&head, loca_format) else {
+        return Err(SubsetTableError(Head::TAG));
+    };
+
+    builder.add_raw(Glyf::TAG, glyf_out);
+    builder.add_raw(Loca::TAG, loca_out);
+    builder.add_raw(Head::TAG, head_out);
+    Ok(())
+}
+
+fn padded_size(len: usize) -> usize {
+    len + len % 2
+}
+
+fn subset_glyph(glyph: &Glyph, plan: &Plan) -> Vec<u8> {
+    //TODO: support set_overlaps_flag and drop_hints
+    match glyph {
+        Composite(comp_g) => subset_composite_glyph(comp_g, plan),
+        Simple(simple_g) => subset_simple_glyph(simple_g),
+    }
+}
+
+// TODO: drop_hints and set_overlaps_flag
+fn subset_simple_glyph(g: &SimpleGlyph) -> Vec<u8> {
+    let mut out = Vec::with_capacity(g.offset_data().len());
+
+    let Some(num_coords) = g.end_pts_of_contours().last() else {
+        return out;
+    };
+    let num_coords = num_coords.get() + 1;
+    let mut coord_bytes: usize = 0;
+    let mut coords_with_flags: u16 = 0;
+
+    let glyph_data = g.glyph_data();
+    let length = glyph_data.len();
+    let mut i: usize = 0;
+    while i < length {
```

Comment:
Can you move the code for trimming padding to a helper function. That will make it clear what this loop is doing and also should help keep this function a bit more manageable once additional hint stripping and overlap flag setting is added.

Reply:
Done for Simple Glyph. For composite glyph, it seems more efficient to do trim padding/hints removal/overlap flag together within one loop.

Reply:
Sounds good.